common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -3774,6 +3775,485 @@ static jint NativeCrypto_EVP_AEAD_CTX_open_buf(JNIEnv* env, jclass, jlong evpAea
+    return ciphertextArray;
+}
+
+static void NativeCrypto_EVP_HPKE_CTX_setup_recipient(JNIEnv* env, jclass, jobject recipientHpkeCtxRef,
+                                                          jobject hpkeKeyCtxRef, jlong evpKdfRef, jlong evpAeadRef,
+                                                          jbyteArray encArray, jbyteArray infoArray) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    EVP_HPKE_CTX* ctx = fromContextObject<EVP_HPKE_CTX>(env, recipientHpkeCtxRef);
+    EVP_HPKE_KEY* key = fromContextObject<EVP_HPKE_KEY>(env, hpkeKeyCtxRef);
+
+    if (ctx == nullptr || key == nullptr) {
+        // NullPointerException thrown while calling fromContextObject
+        return;
+    }
+
+    const EVP_HPKE_KDF* kdf = reinterpret_cast<const EVP_HPKE_KDF*>(evpKdfRef);
+    const EVP_HPKE_AEAD* aead = reinterpret_cast<const EVP_HPKE_AEAD*>(evpAeadRef);
+    if (kdf == nullptr || aead == nullptr) {
+        conscrypt::jniutil::throwNullPointerException(env, "kdf or aead == null");
+        return;
+    }
+
+    if (encArray == nullptr) {
+        conscrypt::jniutil::throwNullPointerException(env, "encArray == null");
+        return;
+    }
+
+    JNI_TRACE("EVP_HPKE_CTX_setup_recipient(%p, %p, %p, %p, %p, %p)",
+               ctx, key, kdf, aead, encArray, infoArray);
+    jbyte* enc = env->GetByteArrayElements(encArray, 0);
+    size_t encLen = env->GetArrayLength(encArray);
+
+    jbyte* infoArrayElement = nullptr;
+    const uint8_t* info = nullptr;
+    size_t infoLen = 0;
+    if (infoArray != nullptr) {
+        infoArrayElement = env->GetByteArrayElements(infoArray, 0);
+        info = reinterpret_cast<const uint8_t*>(infoArrayElement);
+        infoLen = env->GetArrayLength(infoArray);
+    }
+
+    if (!EVP_HPKE_CTX_setup_recipient(/* ctx= */ ctx,
+                                      /* key= */ key,
+                                      /* kdf= */ kdf,
+                                      /* aead= */ aead,
+                                      /* enc= */ reinterpret_cast<const uint8_t*>(enc),
+                                      /* enc_len= */ encLen,
+                                      /* info= */ info,
+                                      /* info_len= */ infoLen)) {
+        env->ReleaseByteArrayElements(encArray, enc, JNI_ABORT);
+        if (infoArrayElement != nullptr) {
+            env->ReleaseByteArrayElements(infoArray, infoArrayElement, JNI_ABORT);
+        }
+        conscrypt::jniutil::throwExceptionFromBoringSSLError(env, "EVP_HPKE_CTX_setup_recipient");
+        return;
+    }
+
+    env->ReleaseByteArrayElements(encArray, enc, JNI_ABORT);
+    if (infoArrayElement != nullptr) {
+        env->ReleaseByteArrayElements(infoArray, infoArrayElement, JNI_ABORT);
+    }
+}
+
+// Defining EVP_HPKE_KEM struct with only the field needed to call the
+// function "EVP_HPKE_CTX_setup_sender_with_seed_for_testing" using "kem->seed_len"
+struct evp_hpke_kem_st {
+  size_t seed_len;
+};
```

Comment:
Please remove this. Reaching into BoringSSL's internal structs like this is not supported. We make structs private for a reason. We can and will change the definition of these struct and won't make any accommodations for projects that do unsupported things like this. In fact, it's already the case that `seed_len` isn't first in the struct, so I'm not sure how this is working at all.

If you need an accessor, tell us and we can add the APIs. But I don't think it's necessary here. Seeds are not part of the official HPKE interface and this is only for testing anyway, it doesn't make sense to need it. Whatever you're getting the test vectors from can be assumed to have a priori knowledge of the seed length the particular KEM wants.

Indeed, since you've got a `seedArray`, it seems you should just be using `env->GetArrayLength(seedArray)`, not `kem->seed_len`, which is actually an out-of-bounds read when `seedArray` didn't happen to be the right size.